test/openssl/test_x509cert.rb

```diff
@@ -276,6 +276,18 @@ def test_marshal
     assert_equal cert.to_der, deserialized.to_der
   end
 
+  def test_load_empty
+    empty_path = Fixtures.file_path("pkey", "empty.pem")
+    certificates = OpenSSL::X509::Certificate.load(empty_path)
+    assert_equal 0, certificates.size
+  end
```

Comment:
Shouldn't it be an error if the file is empty? I think a typical use case of this method is something like this - at least the EE certificate is expected to be included in the file.

```ruby
pkey = OpenSSL::PKey.read(File.read(path))
ee_cert, *ca_certs = OpenSSL::X509::Certificate.load(path)
sslctx = OpenSSL::SSL::SSLContext.new
sslctx.add_certificate(ee_cert, pkey, ca_certs)
```

Reply:
I'm on the fence. Yes, that's main use case, however I also think there is difference between "Garbage file" and "Empty file".

Reply:
cc @tadman wdyt?

Reply:
I checked, and in the above example:

```
a, *b = nil
c, *d = []
```

`a` and `c` are the same, `b` and `d` are the same.

I would say if the file exists and is empty, it's `[]` and if the file doesn't exist, it's exception.

Reply:
To me, an empty file sounds like an invalid input too.

Reply:
Yes, but it's not exceptional. An empty file is common and we can correctly handle it either return `nil` or `[]`. It's a file with zero certificates. I don't believe it represents an exceptional failure.

Reply:
Do you have some examples where an empty file is accepted? I can't come up with anything right now.

While returning an empty array for an empty input file is good for consistency, we've broken it in handling garbage inserted in between PEM blocks. This is probably a good idea, though. A file that contains something but no certificates is likely a result of mixing up files, so silently suppressing that can be a source of hard-to-locate bugs.

```
# Currently:

(input)                     -> (output)
garbage || cert-a || cert-b -> [cert-a, cert-b]
garbage || cert-a           -> [cert-a]
garbage                     -> error

cert-a || cert-b            -> [cert-a, cert-b]
cert-a                      -> [cert-a]
                            -> []
```

I feel the rule is a bit complicated, however. "Raises an error if no certificates could be found" is easier to explain.

Since it's very easy to check if a file is empty or not in Ruby, I think users with special needs should check it beforehand.

Reply:
Thanks for expressing a detailed opinion and giving a set of examples.

I agree, we want to do what makes sense to the user.

Considering empty file an error is one option and I agree, maybe the mental model makes sense, because if I ask myself, is there valid case, I think the answer is no. But for consistency, I like the behaviour either return `nil` (easier to use in `if`) or empty array (more consistent). We certainly had some situations in Ruby where exceptions are used inconsistently:

- `sysread` can raise EOFError. But `read` returns empty string. `read_nonblock` can return nil or sometimes raise exception depending on arguments.
- `Integer(...)` previously raised an exception but it was slow to write `Integer(x) rescue nil`. In this case, it's not "exception" but used to control program flow.

Regarding garbage in the file, my initial reaction was surprise that it would be possible. To me, any garbage should be error! Are you saying that PEM format can contain other text outside the `-----BEGIN/END CERTIFICATE-----` tokens?

To me, if that's the case, file that only contains garbage should return `[]` based on my logic. But I'm not sure I can accept that. Well, it seems odd. To say "Load some certificates" from `/dev/random` and get back an empty array haha.

So, I want to check those cases above first, to see if I can get a better idea of how it should work.

Regarding consistency, in my mind, I imagine there are many use case for this. For example user may have database of files and they want to check if it's certificate or not. Both exception and empty array can be checked, but what makes the most sense?

In particular I don't want to introduce my "opinion" into how something should work. If we chose to return nil, user could write:

```
Certificate.load(path) or raise "Could not load"
```

The question is - if we raise an error for empty file, is it the same kind of error as a file that contains corrupt certificate? I feel it's different.

Ultimately, I'd be sorry to introduce an API where users often write `Certificate.load(path) rescue nil` for typical use case, because it's slow and that's not, IMHO, the point of exceptions. The key here is whether that would be common or not.

Reply:
> Are you saying that PEM format can contain other text outside the `-----BEGIN/END CERTIFICATE-----` tokens?

Unfortunately, yes, and it's common usage. [Apache](https://httpd.apache.org/docs/current/mod/mod_ssl.html#sslcertificatefile) and [Postfix](http://www.postfix.org/postconf.5.html#smtp_tls_chain_files) allow specifying both a certificate (chain) and the corresponding private key from a single file - which contains PEM blocks simply concatenated together.

OpenSSL's `openssl ca` command-line utility also produces a PEM file that contains garbage (a human-readable description of the certificate by `X509_print()`/`OpenSSL::X509::Certificate#to_text`) by default.

> I'd be sorry to introduce an API where users often write `Certificate.load(path) rescue nil` for typical use case, because it's slow and that's not, IMHO, the point of exceptions.

Agreed with that - exceptions should only be used in exceptional conditions. This depends on what expectation we have for this method: to load an arbitrary list of certificates, or to load an EE certificate followed by CA certificates. For the latter, having zero certificates is the same kind of unrecoverable error regardless of the direct cause.

Reply:
Maybe we need high level "convenience" interface and low level "list of certificates".

Reply:
I thought about your examples.

```
# Currently:

(input)                     -> (output)
garbage || cert-a || cert-b -> [cert-a, cert-b]
garbage || cert-a           -> [cert-a]
garbage                     -> error

cert-a || cert-b            -> [cert-a, cert-b]
cert-a                      -> [cert-a]
(empty)                     -> []
```

Since we now support strings, we can simplify our thinking around this, i.e. `(empty)` is `OpenSSL::X509::Certificate.load("")`. We can ignore nature of file system.

I struggle to define `(empty)` case as exceptional. I also see this as different from the `garbage` case because this may represent a corrupt file, which *is* exceptional. Or at least, because we are trying to auto-detect the file type..it's harder for us. For example how would it work if we have the following data:

```
DER encoded cert || PEM encoded cert
```

The current implementation may ignore the initial DER encoded cert. I'm not sure how we should handle this.

For simplicity, I am okay with your proposal to raise an exception. I think it fits the use case pretty well. I'm just concerned that our definition of exceptional in this case might not suit all usage, for example `Integer(..., exception: true/false)` was eventually introduced because so many people doing `Integer(...) rescue nil` which is slow.

Reply:
People can already easily load a single certificate so I don’t think we need to rename this method, and it’s consistent with other gems e.g. yaml.

Reply:
> ```
> DER encoded cert || PEM encoded cert
> ```
>
> The current implementation may ignore the initial DER encoded cert. I'm not sure how we should handle this.

This is somehow correct - it's not a proper DER encoding anyway if there are any extra bytes at the end.

It is possible, however, that DER encoding contains a valid PEM block in the *middle* of the DER-encoded certificate. A certificate can contain arbitrary string and DER-encoding will have the octets as-is. This can be simply fixed by checking DER encoding before PEM encoding.

[ And it seems that the bug exists in `OpenSSL::X509::Certificate.new`, too (and probably in other methods that take both PEM/DER). I will create a PR. ]

But I've started to wonder if we need to support DER within this method at all. Shouldn't the user be able to simply call `OpenSSL::X509::Certificate.new`?